File: python-cfd/Mesh.py
import math
import vtk

class Mesh:
    def __init__(self,dx,dy,h):
        self.O=[0,0]
        self.dim=[int(dx),int(dy)]
        self.size=float(h)
        self.n_points=(dx+1) * (dy+1)
        self.n_cells=dx * dy
        self.h = h

        self.n_cells_x = dx
        self.n_cells_y = dy

        self.n_points_x = dx+1
        self.n_points_y = dy+1

        self.cell_scalars = [0. for _ in range(self.n_cells)]
        self.cell_scalars_name='cell scalar'
        self.cell_vectors=[[0.,0.] for _ in range(self.n_cells)]
        self.cell_vectors_name='cell vector'

        # Point vectors have an x and y component
        self.point_vectors=[[0.,0.] for _ in range(self.n_points)]
        self.point_vectors_name='point vector'
        self.point_scalars=[0. for _ in range(self.n_points)]
        self.point_scalars_name='point scalar'

    ####################
    # Basic
    ####################

    def set_origin(self,x,y):
        self.O=[x,y]

    def index_to_cartesian(self,k,n,nmax):
        if k<0 or k>=nmax:
            return((None,None))
        else:
            (q,r)=divmod(k,n)
            return((r,q))

    def cartesian_to_index(self,i,j,ni,nj):
        if i<0 or i>=ni or j<0 or j>=nj:
            return(None)
        else:
            return(j * ni + i)

    def print(self):
        self.print_meta()
        print("point vectors: {}".format(self.point_vectors))
        print("cell scalars: {}".format(self.cell_scalars))

    def print_meta(self):
        print("origin: ({}; {}) size: {} dimensions: ({}; {})".format(
        self.O[0],self.O[1],self.size,self.dim[0],self.dim[1]))

    def i_max(self): # TODO: rename to i_point_max
        return self.dim[0]

    def j_max(self):
        return self.dim[1]


    ####################
    # Cells
    ####################

    def get_cell_scalars(self):
        return(self.cell_scalars)

    def set_cell_scalars_name(self,name):
        self.cell_scalars_name=name

    def get_cell_scalar(self,i,j):
        k=self.cartesian_to_index(i,j,self.dim[0],self.dim[1])
        if k==None:
            return(math.nan)
        else:
            return(self.cell_scalars[k])

    def set_cell_scalar(self,i,j,scalar):
        k=self.cartesian_to_index(i,j,self.dim[0],self.dim[1])
        if k!=None:
            self.cell_scalars[k]=scalar

    def initialize_cell_scalars_by_coordinate_function(self, fct):
        y = self.O[1] + 0.5 * self.size
        for j in range(self.dim[1]):
            x = self.O[0] + 0.5 * self.size
            for i in range(self.dim[0]):
                print(x,y)
                self.set_cell_scalar(i,j,fct(x,y))
                x += self.size
            y += self.size


    ####################
    # Points
    ####################

    def set_point_vectors_name(self,name):
        self.point_vectors_name=name

    def get_point_vectors(self):
        return(self.point_vectors)

    def get_point_vector(self,i,j):
        k=self.cartesian_to_index(i,j,self.dim[0]+1,self.dim[1]+1)
        if k==None:
            return(math.nan)
        else:
            return(self.point_vectors[k])

    def set_point_vector(self,i,j,c):
        k=self.cartesian_to_index(i,j,self.dim[0]+1,self.dim[1]+1)
        if k!=None:
            self.point_vectors[k]=c

    def set_point_vector_u(self, i, j, u):
        k = self.cartesian_to_index(i, j, self.dim[0]+1, self.dim[1]+1)
        if k != None:
            (self.point_vectors[k])[0] = u

    def set_point_vector_v(self, i, j, v):
        k = self.cartesian_to_index(i, j, self.dim[0]+1, self.dim[1]+1)
        if k != None:
            self.point_vectors[k][1] = v

    def get_point_vector_u(self, i, j):
        return self.get_point_vector(i,j)[0]

    def get_point_vector_v(self, i, j):
        return self.get_point_vector(i,j)[1]

    def initialize_point_vectors_by_coordinate_component_function(self, f):
        y = self.O[1]
        for j in range(self.dim[1]+1):
            x = self.O[0]
            for i in range(self.dim[0]+1):
                self.set_point_vector(i,j,f(x,y))
                x += self.size
            y += self.size


    ####################
    # VTK
    ####################

    def write_vtk(self, file_name):
        ug = vtk.vtkUniformGrid()
        nx=self.dim[0]
        ny=self.dim[1]
        ug.SetDimensions(nx+1, ny+1, 1)
        ug.SetOrigin(self.O[0] ,self.O[0], 0)
        ug.SetSpacing(self.size, self.size, 0)

        # Create cell centered scalar field
        cell_data = vtk.vtkDoubleArray()
        cell_data.SetNumberOfComponents(1)
        cell_data.SetName(self.cell_scalars_name)
        cell_data.SetNumberOfValues(nx * ny)
        for j in range(ny):
            for i in range(nx):
                cell_data.SetValue(j*nx + i, self.get_cell_scalar(i,j))
        ug.GetCellData().SetScalars(cell_data)

        # Create point centered vector field
        point_data = vtk.vtkDoubleArray()
        point_data.SetNumberOfComponents(3)
        point_data.SetName(self.point_vectors_name)
        nxp1 = nx + 1
        nyp1 = ny + 1
        point_data.SetNumberOfTuples(nxp1 * nyp1)
        for j in range(nyp1):
            for i in range(nxp1):
                point_data.SetTuple3(j*nxp1 + i, *(self.get_point_vector(i,j)+[0]))
        ug.GetPointData().SetVectors(point_data)

        writer = vtk.vtkXMLImageDataWriter()
        writer.SetFileName(file_name)
        writer.SetInputData(ug)
        writer.Write()

File: cppfd/mesh.cpp
#include "mesh.h"

#include <iostream>
#include <array>
#include <cmath>
#include <Kokkos_Core.hpp>

#include <vtkDoubleArray.h>
#include <vtkCellData.h>
#include <vtkPointData.h>
#include <vtkUniformGrid.h>
#include <vtkXMLImageDataWriter.h>
#include <vtkXMLImageDataReader.h>

////////////////////////////////////////////////////////////////
// BASIC
////////////////////////////////////////////////////////////////

void Mesh::set_origin(double x, double y){
  this->origin[0] = x;
  this->origin[1] = y;
}

std::array<int,2> Mesh::index_to_cartesian(int k, int n, int nmax){
  if (k < 0 || k >= nmax){
  // Return invalid values when index is out of bounds
    return  {-1, -1};
  } else{
    std::div_t dv = std::div(k, n);
    return {dv.rem, dv.quot};
  }
}

int Mesh::cartesian_to_index(int i, int j, int ni, int nj){
  if(i<0 || i>=ni || j<0 || j>=nj){
    // Return invalid value when coordinates are out of bounds
    return -1;
  } else{
    return j * ni + i;
  }
}

////////////////////////////////////////////////////////////////
// CELLS (PRESSURE)
////////////////////////////////////////////////////////////////

void Mesh::set_pressure(int i, int j, double scalar){
  int k = this->cartesian_to_index(i, j, this->n_cells_x, this->n_cells_y);
  if(k != -1){
    this->pressure(k) = scalar;
  }
}

double Mesh::get_pressure(int i, int j){
  int k = this->cartesian_to_index(i, j, this->n_cells_x, this->n_cells_y);
  if(k == -1){
    return std::nan("");
  } else{
    return this->pressure(k);
  }
}

////////////////////////////////////////////////////////////////
// POINTS (VELOCITY)
////////////////////////////////////////////////////////////////

void Mesh::set_velocity_x(int i, int j, double u){
  if(i > -1 && i < this->get_n_points_x() && j > -1 && j < this->get_n_points_y()){
    this->velocity(i, j, 0) = u;
  }
}

void Mesh::set_velocity_y(int i, int j, double v){
  if(i > -1 && i < this->get_n_points_x() && j > -1 && j < this->get_n_points_y()){
    this->velocity(i, j, 1) = v;
  }
}

double Mesh::get_velocity_x(int i, int j){
  if(i > -1 && i < this->get_n_points_x() && j > -1 && j < this->get_n_points_y()){
    return this->velocity(i, j, 0);
  } else{
    return std::nan("");
  }
}

double Mesh::get_velocity_y(int i, int j){
  if(i > -1 && i < this->get_n_points_x() && j > -1 && j < this->get_n_points_y()){
    return this->velocity(i, j, 1);
  } else{
    return std::nan("");
  }
}

void Mesh::write_vtk(std::string file_name){
  vtkNew<vtkUniformGrid> ug;
  uint64_t nx = this->n_cells_x;
  uint64_t ny = this->n_cells_y;
  ug->SetDimensions(nx + 1, ny + 1, 1);
  ug->SetOrigin(this->origin[0], this->origin[0], 0);
  ug->SetSpacing(this->h, this->h, 0);

  // create cell centered scalar field
  vtkNew<vtkDoubleArray> cell_data;
  cell_data->SetNumberOfComponents(1);
  cell_data->SetName("Pressure");
  cell_data->SetNumberOfValues(nx * ny);
  for(int j = 0; j < ny; j++){
    for(int i = 0; i < nx; i++){
      cell_data->SetValue(j*nx + i, this->get_pressure(i,j));
    }
  }
  ug->GetCellData()->SetScalars(cell_data);

  // create point centered vector field
  vtkNew<vtkDoubleArray> point_data;
  point_data->SetNumberOfComponents(3);
  point_data->SetName("Velocity");
  point_data->SetNumberOfTuples((nx + 1) * (ny + 1));
  for(int j = 0; j < ny +1; j++){
    for(int i = 0; i < nx + 1; i++){
      point_data->SetTuple3(j * (nx + 1) + i, this->get_velocity_x(i, j), this->get_velocity_y(i, j), 0);
    }
  }
  ug->GetPointData()->SetVectors(point_data);

  // write vti visualization file
  vtkNew<vtkXMLImageDataWriter> output_file;
  output_file->SetFileName(file_name.c_str());
  output_file->SetInputData(ug);
  output_file->Write();

  std::cout<<std::endl;
  std::cout<<"Visualization file created: \""<<file_name<<"\""<<std::endl;
}

void Mesh::write_txt(std::string file_name, std::string output_type){
  uint64_t nx = this->n_cells_x;
  uint64_t ny = this->n_cells_y;

  ofstream file;
  file.open(file_name);

  if(!file){
    std::cout << "Error in creating text file" << std::endl;
  }

  if(output_type == "velocity"){
    for(int j = ny; j > - 1; j--){
      std::cout << j << std::endl;
      for(int i = 0; i < nx + 1; i++){
        std::cout << i << std::endl;
        file << this->get_velocity_x(i, j) << " " << this->get_velocity_y(i, j) << "\n";
      }
    }
  }
  else if(output_type == "pressure"){
    for(int j = ny; j > - 1; j--){
      for(int i = 0; i < nx + 1; i++){
        file << this->get_pressure(i, j) << "\n";
      }
    }
  }
  else{
    std::cout << "Unrecognized output type" << '\n';
  }

  //closing the file
  file.close();
  if(file){
    std::cout << std::endl;
    std::cout << "Data file created: \"" << file_name << "\"" << std::endl;
  }
}
